### ISSUE TITLE
Package not installed locally can be accessed by "using", but incorrectly

### ISSUE THREAD

**Issue by zhangcx93** [JuliaLang/julia#55071]
I have a global environment "@v1.10", I have CUDA installed there.
When I'm having a new tesing environment, activated it, having "MLDatasets" added and I forgot to "add CUDA" in that environment.
I can actually run "using CUDA" in such environment.
But CUDA require LLVM, at specific older version, not the latest version, but MLDatasets require LLVM without version specificiation, so if I try to "using CUDA" at this environment, it'll try to load the CUDA with the latest LLVM version, result in a very strange compile error.
It takes me quite a while to figure out what's actually happening.

If a enviroment can access not installed package from global is a feature, it should be done correctly, by respecting the version specification as if it is installed locally.
If it's not a feature, it should notify the user that CUDA is not installed here.

Minimal reproduction:
You have CUDA added in global environment.
```
] add CUDA
```
Then, find a new folder
```
] activate .
] add MLDatasets
using CUDA  
```

**Comment by Seelengrab** [comment]
Isn't this expected behavior & a consequence of how stacked environments work?

**Comment by zhangcx93** [comment]
Yes, after reading the Code Loading doc carefully, I realize it is expected behavior, but not very intuitive.

The idea of [Environment stacks](https://docs.julialang.org/en/v1/manual/code-loading/#Environment-stacks) which override requirements rather than resolving it, is a dangerous but quick solution to provide the feature to load useful tools in global envrionment. And when the conflict happens, it's not obvious where the problem is, the error message always show something deep in other people's package, give the user some strange impression that the code sometimes work and sometimes don't. But it indeed allow the access of global useful tools without introducing much overheads.

How about when loading packages from beyond current envrironment, give an extra warning or info when some error is raised in that package?

**Comment by nhz2** [comment]
Another example of currently confusing loading error messages caused by environment stacks: #55033
